File: hrms/hr/report/employee_attendance_detail/employee_attendance_detail.js
// Copyright (c) 2025, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Employee Attendance Detail"] = {
	filters: [
		{
			fieldname: "employee",
			label: __("员工"),
			fieldtype: "Link",
			options: "Employee",
			reqd: 1,
			get_query: function() {
				return {
					filters: {
						status: "Active"
					}
				};
			}
		},
		{
			fieldname: "year",
			label: __("年份"),
			fieldtype: "Select",
			options: get_year_options(),
			default: new Date().getFullYear(),
			reqd: 1
		},
		{
			fieldname: "month",
			label: __("月份"),
			fieldtype: "Select",
			options: get_month_options(),
			default: new Date().getMonth() + 1,
			reqd: 1
		}
	],
	
	// 颜色含义：
	// 状态：绿色=出勤，红色=缺勤，蓝色=请假，橙色=半天
	// 工时：绿色加粗=≥8小时
	// 迟到/早退：橙色=有标记
	// 星期：灰色=周末
	formatter: function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		if (column.fieldname === "status" && data) {
			if (data.status === "Present") {
				value = `<span style="color: green">${value}</span>`;
			} else if (data.status === "Absent") {
				value = `<span style="color: red">${value}</span>`;
			} else if (data.status === "On Leave") {
				value = `<span style="color: blue">${value}</span>`;
			} else if (data.status === "Half Day") {
				value = `<span style="color: orange">${value}</span>`;
			}
		}
		
		if (column.fieldname === "working_hours" && data && data.working_hours >= 8) {
			value = `<span style="color: green; font-weight: bold">${value}</span>`;
		}
		
		if (column.fieldname === "late_entry" && data && data.late_entry === "✓") {
			value = `<span style="color: orange">${value}</span>`;
		}
		
		if (column.fieldname === "early_exit" && data && data.early_exit === "✓") {
			value = `<span style="color: orange">${value}</span>`;
		}
		
		if (column.fieldname === "day_name" && data) {
			if (data.day_name === "Sat" || data.day_name === "Sun" || 
			    data.day_name === "Saturday" || data.day_name === "Sunday" ||
			    data.day_name === "周六" || data.day_name === "周日") {
				value = `<span style="color: #999">${value}</span>`;
			}
		}
		
		return value;
	}
};

function get_year_options() {
	let options = [];
	let current_year = new Date().getFullYear();
	for (let i = current_year; i >= current_year - 5; i--) {
		options.push(i);
	}
	return options;
}

function get_month_options() {
	return [
		{ value: 1, label: __("1月") },
		{ value: 2, label: __("2月") },
		{ value: 3, label: __("3月") },
		{ value: 4, label: __("4月") },
		{ value: 5, label: __("5月") },
		{ value: 6, label: __("6月") },
		{ value: 7, label: __("7月") },
		{ value: 8, label: __("8月") },
		{ value: 9, label: __("9月") },
		{ value: 10, label: __("10月") },
		{ value: 11, label: __("11月") },
		{ value: 12, label: __("12月") },
	];
}

File: hrms/hr/report/employee_attendance_detail/employee_attendance_detail.py
# Copyright (c) 2025, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate, formatdate
from calendar import monthrange


def execute(filters=None):
    filters = frappe._dict(filters or {})
    
    if not filters.get("employee"):
        frappe.msgprint(_("Please select an employee"), alert=True)
        return [], []
    
    columns = get_columns(filters)
    data = get_data(filters)
    summary = get_summary(data)
    # 不显示图表，让表格更清晰
    return columns, data, summary, None


def get_columns(filters):
    return [
        {
            "label": _("日期"),
            "fieldname": "attendance_date",
            "fieldtype": "Date",
            "width": 120,
        },
        {
            "label": _("星期"),
            "fieldname": "day_name",
            "fieldtype": "Data",
            "width": 80,
        },
        {
            "label": _("状态"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": _("签到"),
            "fieldname": "in_time",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": _("签退"),
            "fieldname": "out_time",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": _("工时"),
            "fieldname": "working_hours",
            "fieldtype": "Float",
            "precision": 2,
            "width": 100,
        },
        {
            "label": _("迟到"),
            "fieldname": "late_entry",
            "fieldtype": "Data",
            "width": 60,
        },
        {
            "label": _("早退"),
            "fieldname": "early_exit",
            "fieldtype": "Data",
            "width": 60,
        },
        {
            "label": _("班次"),
            "fieldname": "shift",
            "fieldtype": "Link",
            "options": "Shift Type",
            "width": 100,
        },
    ]


def get_data(filters):
    conditions = get_conditions(filters)
    
    data = frappe.db.sql(
        """
        SELECT 
            a.attendance_date,
            DAYNAME(a.attendance_date) as day_name,
            a.status,
            TIME(a.in_time) as in_time,
            TIME(a.out_time) as out_time,
            a.working_hours,
            CASE WHEN a.late_entry = 1 THEN '✓' ELSE '' END as late_entry,
            CASE WHEN a.early_exit = 1 THEN '✓' ELSE '' END as early_exit,
            a.shift
        FROM `tabAttendance` a
        WHERE a.docstatus = 1
          AND a.employee = %(employee)s
          {conditions}
        ORDER BY a.attendance_date
        """.format(conditions=conditions),
        filters,
        as_dict=True
    )
    
    # Format time fields
    for row in data:
        if row.in_time:
            row.in_time = str(row.in_time)[:5]  # HH:MM
        else:
            row.in_time = "-"
        if row.out_time:
            row.out_time = str(row.out_time)[:5]  # HH:MM
        else:
            row.out_time = "-"
        
        # Translate day name
        day_map = {
            "Monday": _("Mon"),
            "Tuesday": _("Tue"),
            "Wednesday": _("Wed"),
            "Thursday": _("Thu"),
            "Friday": _("Fri"),
            "Saturday": _("Sat"),
            "Sunday": _("Sun"),
        }
        row.day_name = day_map.get(row.day_name, row.day_name)
    
    return data


def get_conditions(filters):
    conditions = []
    
    if filters.get("year") and filters.get("month"):
        year = int(filters.get("year"))
        month = int(filters.get("month"))
        _, last_day = monthrange(year, month)
        start_date = f"{year}-{month:02d}-01"
        end_date = f"{year}-{month:02d}-{last_day}"
        conditions.append("AND a.attendance_date BETWEEN %(start_date)s AND %(end_date)s")
        filters["start_date"] = start_date
        filters["end_date"] = end_date
    
    return " ".join(conditions)


def get_summary(data):
    if not data:
        return []
    
    present_days = sum(1 for d in data if d.status == "Present")
    absent_days = sum(1 for d in data if d.status == "Absent")
    leave_days = sum(1 for d in data if d.status == "On Leave")
    half_days = sum(1 for d in data if d.status == "Half Day")
    total_hours = sum(d.working_hours or 0 for d in data)
    late_entries = sum(1 for d in data if d.late_entry == "✓")
    early_exits = sum(1 for d in data if d.early_exit == "✓")
    
    return [
        {
            "value": present_days,
            "label": _("Present Days"),
            "indicator": "green",
            "datatype": "Int",
        },
        {
            "value": absent_days,
            "label": _("Absent Days"),
            "indicator": "red",
            "datatype": "Int",
        },
        {
            "value": leave_days,
            "label": _("Leave Days"),
            "indicator": "blue",
            "datatype": "Int",
        },
        {
            "value": round(total_hours, 2),
            "label": _("Total Working Hours"),
            "indicator": "orange",
            "datatype": "Float",
        },
        {
            "value": late_entries,
            "label": _("Late Entries"),
            "indicator": "yellow",
            "datatype": "Int",
        },
    ]


def get_chart(data):
    if not data:
        return None
    
    # Daily working hours chart
    dates = [str(d.attendance_date.day) for d in data if d.working_hours]
    hours = [d.working_hours for d in data if d.working_hours]
    
    if not dates:
        return None
    
    return {
        "data": {
            "labels": dates,
            "datasets": [
                {
                    "name": _("Working Hours"),
                    "values": hours,
                }
            ],
        },
        "type": "bar",
        "colors": ["#7cd6fd"],
        "height": 200,
    }

